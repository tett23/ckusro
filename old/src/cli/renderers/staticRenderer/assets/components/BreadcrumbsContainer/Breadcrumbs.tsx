import React from 'react';
import styled from 'styled-components';
import { FileBuffer } from '../../../../../../models/FileBuffer';
import BreadcrumbItem from './BreadcrumbItem';
import getAncestors from './getAncesters';

export type Props = {
  fileBuffers: FileBuffer[];
  fileBuffer: FileBuffer;
};

export default function Breadcrumbs({ fileBuffers, fileBuffer }: Props) {
  const items = [fileBuffer]
    .concat(getAncestors(fileBuffer, fileBuffers))
    .map((item) => {
      return <BreadcrumbItem key={item.id} fileBuffer={item} />;
    });

  return (
    <nav>
      <Ul>{items}</Ul>
    </nav>
  );
}

const Ul = styled.ul`
  display: flex;
  flex-direction: row;
  list-style: none;
  padding: 0;
  margin: 0;
`;
