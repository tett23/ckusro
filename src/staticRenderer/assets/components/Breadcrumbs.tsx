import { join as pathJoin, normalize, sep } from 'path';
import React from 'react';

export type Props = {
  namespace: string;
  path: string;
};

export default function Breadcrumbs({ namespace, path }: Props) {
  const items = pathItems(namespace, path).map(([text, itemPath]) => (
    <BreadcrumbItem key={itemPath} text={text} path={itemPath} />
  ));

  return (
    <nav>
      <ul>{items}</ul>
    </nav>
  );
}

export function pathItems(
  namespace: string,
  path: string,
): Array<[string, string]> {
  return splitItems(namespace, path)
    .reverse()
    .map(
      (item, i): [string, string] => {
        const itemPath = pathJoin(
          './',
          ...Array.from({ length: i }, () => '').fill('..'),
        );

        return [item, itemPath];
      },
    )
    .reverse();
}

function splitItems(namespace: string, path: string): string[] {
  const items = normalize(path)
    .split(sep)
    .filter((item) => item !== '');
  const ret = [namespace, items].flatMap((item) => item);

  return ret;
}

export type BreadcrumbItemProps = {
  text: string;
  path: string;
};

export function BreadcrumbItem({ text, path }: BreadcrumbItemProps) {
  return (
    <li>
      <a href={path}>{text}</a>
    </li>
  );
}
