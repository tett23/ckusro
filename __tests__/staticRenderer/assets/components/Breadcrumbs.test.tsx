import { shallow } from 'enzyme';
import React from 'react';
import Breadcrumbs, {
  pathItems,
} from '../../../../src/staticRenderer/assets/components/Breadcrumbs';

describe(Breadcrumbs, () => {
  it('renders correctly', () => {
    const wrapper = shallow(
      <Breadcrumbs namespace="namespace" path="/foo.md" />,
    );

    expect(wrapper).toMatchSnapshot();
  });
});

describe(pathItems, () => {
  it('returns name-path tuple', () => {
    const data: Array<[string, Array<[string, string]>]> = [
      [
        '/foo/bar/baz.md',
        [
          ['namespace', '../../..'],
          ['foo', '../..'],
          ['bar', '..'],
          ['baz.md', './'],
        ],
      ],
      ['/foo.md', [['namespace', '..'], ['foo.md', './']]],
      ['/', [['namespace', './']]],
    ];
    data.forEach(([path, expected]) => {
      const actual = pathItems('namespace', path);

      expect(actual).toEqual(expected);
    });
  });
});
