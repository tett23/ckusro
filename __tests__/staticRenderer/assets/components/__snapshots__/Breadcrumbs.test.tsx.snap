// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumbs renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Breadcrumbs
    namespace="namespace"
    path="/foo.md"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <ul>
        <BreadcrumbItem
          path=".."
          text="namespace"
        />
        <BreadcrumbItem
          path="./"
          text="foo.md"
        />
      </ul>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <BreadcrumbItem
            path=".."
            text="namespace"
          />,
          <BreadcrumbItem
            path="./"
            text="foo.md"
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "..",
          "nodeType": "function",
          "props": Object {
            "path": "..",
            "text": "namespace",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "./",
          "nodeType": "function",
          "props": Object {
            "path": "./",
            "text": "foo.md",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "ul",
    },
    "type": "nav",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <ul>
          <BreadcrumbItem
            path=".."
            text="namespace"
          />
          <BreadcrumbItem
            path="./"
            text="foo.md"
          />
        </ul>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <BreadcrumbItem
              path=".."
              text="namespace"
            />,
            <BreadcrumbItem
              path="./"
              text="foo.md"
            />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "..",
            "nodeType": "function",
            "props": Object {
              "path": "..",
              "text": "namespace",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "./",
            "nodeType": "function",
            "props": Object {
              "path": "./",
              "text": "foo.md",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "ul",
      },
      "type": "nav",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
